core/src/toga/images.py

```diff
@@ -1,54 +1,91 @@
+        :param src: The source from which to load the image. Can be a file path
+            (relative or absolute, as a string or :external:class:`pathlib.Path`) or raw
+            binary data in any supported image format. Can also accept a
+            :external:class:`PIL.Image.Image` if Pillow is installed.
+        :param path: **DEPRECATED** - Use ``src``.
+        :param data: **DEPRECATED** - Use ``src``.
         :raises FileNotFoundError: If a path is provided, but that path does not exist.
-        :raises ValueError: If the path or data cannot be loaded as an image.
+        :raises ValueError: If the source cannot be loaded as an image.
         """
-        if path is None and data is None:
-            raise ValueError("Either path or data must be set.")
-        if path is not None and data is not None:
-            raise ValueError("Only either path or data can be set.")
-
+        ######################################################################
+        # 2023-11: Backwards compatibility
+        ######################################################################
+        num_provided = sum(arg is not None for arg in (src, path, data))
+        if num_provided > 1:
+            raise ValueError("Received multiple arguments to constructor.")
+        if num_provided == 0:
+            raise ValueError("No image source supplied.")
         if path is not None:
-            if isinstance(path, Path):
-                self.path = path
-            else:
-                self.path = Path(path)
-        else:
-            self.path = None
+            src = path
+            warn(
+                "Path argument is deprecated, use src instead.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        elif data is not None:
+            src = data
+            warn(
+                "Data argument is deprecated, use src instead.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        ######################################################################
+        # End backwards compatibility
+        ######################################################################
 
         self.factory = get_platform_factory()
-        if data is not None:
-            self._impl = self.factory.Image(interface=self, data=data)
+        self._path = None
+
+        # Any "lump of bytes" should be valid here.
+        if isinstance(src, (bytes, bytearray, memoryview)):
+            self._impl = self.factory.Image(interface=self, data=src)
+
+        elif isinstance(src, (str, Path)):
+            self._path = toga.App.app.paths.app / src
+            if not self._path.is_file():
+                raise FileNotFoundError(f"Image file {self._path} does not exist")
+            self._impl = self.factory.Image(interface=self, path=self._path)
+
+        elif isinstance(src, PIL_Image.Image):
```

Comment:
For the sake of completeness - I wonder if we should include Toga.Image in this list. It's a dumb use case, but it would make the type declarations symmetrical, and I can imagine there might be uses where a completely generic pipeline might want to round trip an image into itself.

Reply:
I hadn't thought of that but you're right, that's not a bad idea.